packages/modules/psa/api.py

```diff
@@ -0,0 +1,155 @@
+        realm = 'clientsB2CDS'
+        if (not client_id) or (not client_secret):
+            client_id = "cbf74ee7-a303-4c3d-aba3-29f5994e2dfa"
+            client_secret = "X6bE6yQ3tH1cG5oA6aW4fS6hK0cR0aK5yN2wE4hP8vL8oW5gU3"
+    elif (manufacturer == "Opel"):
+        brand = 'opel.com'
+        realm = 'clientsB2COpel'
+        if (not client_id) or (not client_secret):
+            client_id = "07364655-93cb-4194-8158-6b035ac2c24c"
+            client_secret = "F2kK7lC5kF5qN7tM0wT8kE3cW1dP0wC5pI6vC0sQ5iP5cN8cJ8"
+    elif (manufacturer == "Vauxhall"):
+        brand = 'vauxhall.co.uk'
+        realm = 'clientsB2CVauxhall'
+        # don't have valid client_id and secret for Vauxhall
+
+    app_credentials = client_id + ':' + client_secret
+    encoded_app_credentials = base64.b64encode(app_credentials.encode()).decode()
+
+    data = {'realm': realm, 'grant_type': 'password', 'password': password, 'username': user_id, 'scope': scope}
+    headers = {
+        'Content-Type': 'application/x-www-form-urlencoded', 'Authorization': 'Basic %s' % encoded_app_credentials}
+    psa_config = req.get_http_session().post(
+        f'https://idpcvs.{brand}/am/oauth2/access_token', data=data, headers=headers).json()
+
+    access_token = psa_config['access_token']
+
+    # Step 2: Retrieve VINs and VIN_IDs in profile
+    data = {'client_id': client_id}
+    headers = {
+        'Accept': 'application/hal+json', 'Authorization': 'Bearer %s' % access_token, 'x-introspect-realm': realm}
+
+    vehicle_response = req.get_http_session().get(
+        f'https://api.groupe-psa.com/connectedcar/v4/user/vehicles?client_id={client_id}', headers=headers).json()
+    vehicles = vehicle_response['_embedded']['vehicles']
+
+    # Filter list for given VIN or select first entry if no VIN is provided
+    try:
+        vehicle_selected = next(
+            vehicle for vehicle in vehicles if vehicle['vin'] == vehicle_vin) if vehicle_vin else vehicles[0]
+        vehicle_id = vehicle_selected['id']
+        vehicle_vin_found = vehicle_selected['vin']
+    except Exception:
+        log.error(f'Cannot find VIN {vehicle_vin} in PSA account {user_id}. Check VIN and retry.')
+        return
+
+    log.debug(f'Fetching details for VIN: {vehicle_vin_found} with vehicle_id:{vehicle_id}')
+
+    # Step 3: get details for selected vehicle
+    data = {'client_id': client_id}
+    headers = {
+        'Accept': 'application/hal+json', 'Authorization': 'Bearer %s' % access_token, 'x-introspect-realm': realm}
+    battery = req.get_http_session().get(
+        f'https://api.groupe-psa.com/connectedcar/v4/user/vehicles/{vehicle_id}/status?client_id={client_id}',
+        headers=headers).json()
+
+    # filter to only include type=Electric but remove all others. Seen type=Fuel and type=Electric being returned.
+    battery = list(filter(lambda x: x['type'] == 'Electric', battery['energy']))
+    soc = battery[0]['level']
+    range = battery[0]['autonomy']
```

Comment:
Wir sollten hier auch den neuen Code verwenden:

```suggestion
    energies = battery['energy']
    energy_selected = next(energy for energy in energies if energy['type'] == 'Electric')
    soc = energy_selected['level']
    range = energy_selected['autonomy']
    fetchedsoctime = energy_selected['updatedAt']
```

Reply:
ah ja danke, das hatte ich beim manuellen Übernehemen des Filters gar nicht gesehen. Ist nun mit drin.